File: 프로그래머스/2/12952. N－Queen/N－Queen.cs
using System;

public class Solution {
    static int ans = 0;
    static int[] arr = new int[12];

    static bool check(int level)
    {
        for (int i = 0; i < level; i++)
        {
            if (arr[i] == arr[level] || Math.Abs(arr[i] - arr[level]) == level - i) return false;
        }
        return true;
    }
    
    static void nqueen(int n, int num)
    {
        if (num == n) ans++;
        else
        {
            for (int i = 0; i < n; i++)
            {
                arr[num] = i;
                if (check(num)) nqueen(n, num + 1);
            }
        }
    }
    
    public int solution(int n) {
        nqueen(n, 0);
        return ans;
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
using namespace std;

int n, ans = 0;
int arr[15];

bool check(int level) {
	for (int i = 0; i < level; i++) {
		if (arr[i] == arr[level] || abs(arr[i] - arr[level]) == level - i) return false;
	}
	return true;
}

void nqueen(int num) {
	if (num == n) ans++;
	else {
		for (int i = 0; i < n; i++) {
			arr[num] = i;
			if (check(num)) nqueen(num + 1);
		}
	}
}

int main () {
	ios_base :: sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	cin >> n;
	nqueen(0);
	cout << ans;
	return 0;
}